File: dex/code_block.py
import enum
import logging
from dataclasses import dataclass
from typing import Optional

from dex.helpers import sign_extend
from dex.instructions_new import InstructionBase
from vm.utils import LogHandler

handler = LogHandler()
log = logging.getLogger(__name__)
log.addHandler(handler)
log.setLevel(logging.DEBUG)

class BasicBlockType(enum.Enum):
    RETURN  = 0
    THROW   = 1
    GOTO    = 2
    IF      = 3
    SWITCH  = 4
    GENERIC = 5

class EdgeKind(enum.Enum):
    FALL_THROUGH = 0
    JUMP         = 1
    EXCEPTION    = 2
    SWITCH       = 3

class CFGEdge:
    def __init__(self,
                 source: 'BasicBlock',
                 target: 'BasicBlock',
                 kind: EdgeKind,
                 switch_key: Optional[int] = None):

        self.source = source
        self.target = target
        self.kind   = kind
        self.switch_key = switch_key


class BasicBlock:
    id: int

    def __init__(self):
        # For the VM
        self.instructions: list = []
        self.next_branch: Optional[int] = None # (If it branches off)
        self.block_type: BasicBlockType = BasicBlockType.GENERIC
        self.instr_idx_start = 0

        # CFG
        self.id = 0
        self.successors:   list[CFGEdge] = []
        self.predecessors: list[CFGEdge] = []

        # Dominator fields
        self.dominator: Optional['BasicBlock'] = None
        self.dom_children: list = []
        self.dom_frontier: list = []

        self.loop_header: bool      = False
        self.loop: Optional[object] = None
        self.ssa_defs: dict         = {}

    @property
    def first_codepoint(self) -> int:
        return self.instructions[0].codepoint if self.instructions else -1

    @property
    def last_instruction(self) -> int:
        return self.instructions[-1] if self.instructions else None

    def add_successor(self, target: 'BasicBlock', kind: EdgeKind, switch_key: Optional[int] = None):
        edge = CFGEdge(self, target, kind, switch_key)
        self.successors.append(edge)
        target.predecessors.append(edge)

class CodeBlock:
    """
        This implementation of code block is not synonymous with the documentation. Instead, it
        intends to put most of the base implementation in the Method class with the breaking out of
        instructions and control flow (Goto, TryCatch, etc) into blocks here.

        This will contain all control flow related "blocks" that can then be parsed later or used to
        build a callgraph

        'code_item' Reference: https://source.android.com/docs/core/runtime/dex-format#code-item
    """

    def __init__(self, code_item):
        self.blocks: list[BasicBlock] = []
        self.code_item = code_item
        self.addr_lookup: dict[int, BasicBlock] = {}
        self._block_id: int = 0

    @property
    def entry(self) -> Optional[BasicBlock]:
        return self.blocks[0] if self.blocks else None

    def build_code_flow(self):
        # 'instr_size' should cover this, but just to be safe, we'll do an extra check
        instrs = self.code_item.instructions

        if not instrs:
            return

        leaders = self._find_leaders(instrs)
        self._build_blocks(instrs, leaders)
        self._connect_edges()
        self._add_exception_edges()


    # "leader" == first instruction of a block
    def _find_leaders(self, instrs):
        leaders = {instrs[0].codepoint}

        for i, instr in enumerate(instrs):
            op = instr.opcode

            def mark_next():
                # make sure we don't accidentally pass out of list of instructions
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)

            match op:
                case 0x28:
                    leaders.add(instr.codepoint + sign_extend(instr.vA, 8))
                    mark_next()
                case 0x29:
                    leaders.add(instr.codepoint + sign_extend(instr.vA, 16))
                    mark_next()
                case 0x2a:
                    leaders.add(instr.codepoint + sign_extend(instr.vA, 32))
                    mark_next()
                case _ if 0x32 <= op <= 0x37:
                    leaders.add(instr.codepoint + sign_extend(instr.vA, 16))
                    mark_next()
                case _ if 0x38 <= op <= 0x3d:
                    leaders.add(instr.codepoint + sign_extend(instr.vA, 16))
                    mark_next()
                case _ if op in (0x2b, 0x2c):
                    for rel in instr.switch_table.values():
                        leaders.add(instr.codepoint + rel)
                    mark_next()
                case _ if op in (0x0e, 0x0f, 0x10, 0x11, 0x27):
                    mark_next()

            return leaders


    def _build_blocks(self, instrs, leaders):
        current = None
        for instr in instrs:
            if instr.codepoint in leaders:
                current = BasicBlock()
                current.id = self._block_id
                current.instr_idx_start = len(self.blocks)
                self._block_id += 1
                self.blocks.append(current)
                self.addr_lookup[instr.codepoint] = current

            if current is not None:
                current.instructions.append(instr)

        for block in self.blocks:
            self._classify_block(block)


    def _classify_block(self, block):
        if not block.instructions:
            return
        last = block.last_instruction
        op   = last.opcode

        match op:
            case _ if 0x0e <= op <= 0x11:
                block.block_type = BasicBlockType.RETURN
            case 0x27:
                block.block_type = BasicBlockType.THROW
            case 0x28:
                block.block_type = BasicBlockType.GOTO
                block.next_branch = last.codepoint + sign_extend(last.vA, 8)
            case 0x29:
                block.block_type = BasicBlockType.GOTO
                block.next_branch = last.codepoint + sign_extend(last.vA, 16)
            case 0x2a:
                block.block_type = BasicBlockType.GOTO
                block.next_branch = last.codepoint + sign_extend(last.vA, 32)
            case _ if 0x32 <= op <= 0x37:
                block.block_type = BasicBlockType.IF
                block.next_branch = last.codepoint + sign_extend(last.vC, 16)
            case _ if 0x38 <= op <= 0x3d:
                block.block_type = BasicBlockType.IF
                block.next_branch = last.codepoint + sign_extend(last.vB, 16)
            case _ if op in (0x2b, 0x2c):
                block.block_type = BasicBlockType.SWITCH
            case _:
                block.block_type = BasicBlockType.GENERIC

    def _connect_edges(self):
        sorted_cps = sorted(self.addr_lookup.keys())
        for block in self.blocks:
            if not block.instructions:
                continue

            last = block.last_instruction
            op = last.opcode

            def next_block():
                idx = sorted_cps.index(block.first_codepoint)
                if idx + 1 < len(sorted_cps):
                    return self.addr_lookup[sorted_cps[idx + 1]]
                return None

            def target(codepoint):
                return self.addr_lookup.get(codepoint)

            match op:
                case 0x28:
                    t = target(last.codepoint + sign_extend(last.vA, 8))
                    if t:
                        block.add_successor(t, EdgeKind.JUMP)
                case 0x29:
                    t = target(last.codepoint + sign_extend(last.vA, 16))
                    if t:
                        block.add_successor(t, EdgeKind.JUMP)
                case 0x28:
                    t = target(last.codepoint + sign_extend(last.vA, 32))
                    if t:
                        block.add_successor(t, EdgeKind.JUMP)

                case _ if 0x32 <= op <= 0x37:
                    true_target = target(last.codepoint + sign_extend(last.vC, 16))
                    false_target = next_block()
                    if true_target:
                        block.add_successor(true_target, EdgeKind.JUMP)
                    if false_target:
                        block.add_successor(false_target, EdgeKind.FALL_THROUGH)

                case _ if 0x38 <= op <= 0x3d:
                    true_target = target(last.codepoint + sign_extend(last.vB, 16))
                    false_target = next_block()
                    if true_target:
                        block.add_successor(true_target, EdgeKind.JUMP)
                    if false_target:
                        block.add_successor(false_target, EdgeKind.FALL_THROUGH)

                case _ if op in (0x2b, 0x2c):
                    for key, rel in last.switch_table.items():
                        t = target(last.codepoint + rel)
                        if t:
                            block.add_successor(t, EdgeKind.SWITCH, switch_key=key)
                    next_blk = next_block()
                    if next_blk:
                        block.add_successor(next_blk, EdgeKind.FALL_THROUGH)

                case _ if op in (0x0e, 0x0f, 0x10, 0x11, 0x27):
                    pass # terminators -> they don't have successors

                case _:
                    next_blk = next_block()
                    if next_blk:
                        block.add_successor(next_blk, EdgeKind.FALL_THROUGH)

    def _add_exception_edges(self):
        for try_item in self.code_item.try_items:
            start = try_item.start_addr
            end   = start + try_item.insn_count

            for cp, block in self.addr_lookup.items():
                if not (start <= cp < end):
                    continue

                for handler in self.code_item.handlers:
                    for h in handler.handlers:
                        handler_block = self.addr_lookup[h['addr']]
                        if handler_block:
                            block.add_successor(handler_block, EdgeKind.EXCEPTION)
                    if handler.catch_all_addr:
                        handler_block = self.addr_lookup.get(handler.catch_all_addr)
                        if handler_block:
                            block.add_successor(handler_block, EdgeKind.EXCEPTION)

    def lookup_block_by_codepoint(self, codepoint):
        return self.addr_lookup.get(codepoint)

    def lookup_codeblock_by_idx_offset(self, idx):
        if 0 <= idx < len(self.blocks):
            return self.blocks[idx]
        return None

    def reverse_postorder(self):
        if not self.blocks:
            return []
        visited = set()
        result  = []

        def dfs(block: BasicBlock):
            if block in visited:
                return

            visited.add(block)
            for edge in block.successors:
                if edge.kind != EdgeKind.EXCEPTION:
                    dfs(edge.target)
            result.append(block)

        dfs(self.blocks[0])
        result.reverse()
        return result


# control flow graphb -- builder
class CFGBuilder:

    def __init__(self, method):
        self.method = method
        self.instrs = method.instructions
        self.blocks = {}  # codepoint to block
        self._block_id = 0

    def build(self):
        leaders = self._find_leaders()
        self._build_blocks(leaders)
        self._connect_blocks()
        self._add_exception_edges()
        return CFG(
            entry = self.blocks[0],
            blocks = list(self.blocks.values()),
            method = self.method,
        )


    def _find_leaders(self):
        leaders = {0}
        instrs = self.instrs

        for i, instr in enumerate(instrs):
            op = instr.opcode

            # unconditional branches
            if op in (0x28, 0x29, 0x2a):
                bits = {0x28: 8, 0x29: 16, 0x2a: 32}[op]
                target = instr.codepoint + sign_extend(instr.vA, bits)
                leaders.add(target)
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)

            # conditional branches
            elif 0x32 <= op <= 0x37:
                target = instr.codepoint + sign_extend(instr.vC, 16)
                leaders.add(target)
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)

            elif op == 0x38 <= op <= 0x3d:
                target = instr.codepoint + sign_extend(instr.vB, 16)
                leaders.add(target)
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)

            # Switch
            elif op in (0x2b, 0x2c):
                for rel in instr.switch_table.values():
                    leaders.add(instr.codepoint + rel)
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)

            # return / throw (end of block)
            elif 0x0e <= op <= 0x11 or op == 0x27:
                if i + 1 < len(instrs):
                    leaders.add(instrs[i + 1].codepoint)


        return leaders


    def _build_blocks(self, leaders):
        current_block = None
        for instr in self.instrs:
            if instr.codepoint in leaders:
                current_block = BasicBlock(id=self._block_id)
                self._block_id += 1
                self.blocks[instr.codepoint] = current_block

            if current_block is not None:
                current_block.instructions.append(instr)


@dataclass
class CFG:
    entry: BasicBlock
    blocks: list
    method: object

    def reverse_postorder(self):
        visited = set()
        result = []

        def dfs(block):
            if block in visited:
                return
            visited.add(block)

            for edge in block.successors:
                dfs(edge.target)
            result.append(block)

        dfs(self.entry)
        result.reverse()
        return result

File: experimental/rust/platypus-dex/src/code_block.rs
/// Basic block / CFG construction — translates dex/code_block.py

use std::collections::{HashMap, HashSet};

use super::instructions::{Instruction, InstructionKind};
use super::helpers::sign_extend;

// ── Block & edge types ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BasicBlockType {
    Return,
    Throw,
    Goto,
    If,
    Switch,
    Generic,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EdgeKind {
    FallThrough,
    Jump,
    Exception,
    Switch,
}

#[derive(Debug, Clone)]
pub struct CfgEdge {
    pub source_id: usize,
    pub target_id: usize,
    pub kind: EdgeKind,
    pub switch_key: Option<i32>,
}

#[derive(Debug, Clone)]
pub struct BasicBlock {
    pub id: usize,
    /// Indices into the instruction vector.
    pub instr_indices: Vec<usize>,
    pub block_type: BasicBlockType,
    /// Codepoint of the first instruction (or u32::MAX if empty).
    pub first_codepoint: u32,
    /// Absolute codepoint target for GOTO/IF blocks.
    pub next_branch: Option<u32>,

    // CFG connectivity (stored as edge indices into the parent CFG's edge vec)
    pub successor_edges:   Vec<usize>,
    pub predecessor_edges: Vec<usize>,

    // Dominator fields (populated by a separate pass if needed)
    pub dominator: Option<usize>,
    pub dom_children: Vec<usize>,
    pub dom_frontier: Vec<usize>,

    pub loop_header: bool,
}

impl BasicBlock {
    fn new(id: usize) -> Self {
        BasicBlock {
            id,
            instr_indices: Vec::new(),
            block_type: BasicBlockType::Generic,
            first_codepoint: u32::MAX,
            next_branch: None,
            successor_edges: Vec::new(),
            predecessor_edges: Vec::new(),
            dominator: None,
            dom_children: Vec::new(),
            dom_frontier: Vec::new(),
            loop_header: false,
        }
    }
}

// ── CFG ───────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct Cfg {
    pub blocks: Vec<BasicBlock>,
    pub edges: Vec<CfgEdge>,
    /// Maps codepoint → block id.
    pub addr_lookup: HashMap<u32, usize>,
}

impl Cfg {
    /// Reverse post-order traversal (excluding exception edges).
    pub fn reverse_postorder(&self) -> Vec<usize> {
        if self.blocks.is_empty() {
            return Vec::new();
        }
        let mut visited = HashSet::new();
        let mut result  = Vec::new();
        self.dfs_rpo(0, &mut visited, &mut result);
        result.reverse();
        result
    }

    fn dfs_rpo(&self, block_id: usize, visited: &mut HashSet<usize>, result: &mut Vec<usize>) {
        if visited.contains(&block_id) {
            return;
        }
        visited.insert(block_id);
        let block = &self.blocks[block_id];
        for &edge_idx in &block.successor_edges {
            let edge = &self.edges[edge_idx];
            if edge.kind != EdgeKind::Exception {
                self.dfs_rpo(edge.target_id, visited, result);
            }
        }
        result.push(block_id);
    }

    /// Look up a block by codepoint.
    pub fn block_at(&self, codepoint: u32) -> Option<&BasicBlock> {
        self.addr_lookup.get(&codepoint).map(|&id| &self.blocks[id])
    }
}

// ── Builder ───────────────────────────────────────────────────────────────────

pub struct CfgBuilder<'a> {
    instructions: &'a [Instruction],
}

impl<'a> CfgBuilder<'a> {
    pub fn new(instructions: &'a [Instruction]) -> Self {
        CfgBuilder { instructions }
    }

    pub fn build(
        &self,
        try_items: &[super::parser::TryItem],
        handlers: &[super::parser::EncodedCatchHandler],
    ) -> Cfg {
        let leaders = self.find_leaders();
        let (mut blocks, addr_lookup, instr_to_block) = self.build_blocks(&leaders);

        for block in &mut blocks {
            classify_block(block, self.instructions);
        }

        let mut edges: Vec<CfgEdge> = Vec::new();
        let sorted_cps: Vec<u32> = {
            let mut v: Vec<u32> = addr_lookup.keys().copied().collect();
            v.sort_unstable();
            v
        };

        connect_edges(&mut blocks, &mut edges, &addr_lookup, &sorted_cps, self.instructions);
        add_exception_edges(&mut blocks, &mut edges, &addr_lookup, try_items, handlers);

        Cfg { blocks, edges, addr_lookup }
    }

    fn find_leaders(&self) -> HashSet<u32> {
        let instrs = self.instructions;
        if instrs.is_empty() {
            return HashSet::new();
        }
        let mut leaders: HashSet<u32> = HashSet::new();
        leaders.insert(instrs[0].codepoint);

        for (i, instr) in instrs.iter().enumerate() {
            let op = instr.opcode;

            let mark_next = |leaders: &mut HashSet<u32>| {
                if i + 1 < instrs.len() {
                    leaders.insert(instrs[i + 1].codepoint);
                }
            };

            match op {
                // unconditional goto family
                0x28 | 0x29 | 0x2a => {
                    let bits = match op { 0x28 => 8, 0x29 => 16, _ => 32 };
                    let offset = instr.v_a.unwrap_or(0);
                    let target = (instr.codepoint as i64 + sign_extend(offset, bits)) as u32;
                    leaders.insert(target);
                    mark_next(&mut leaders);
                }
                // if-* (two-register)
                0x32..=0x37 => {
                    let offset = instr.v_c.unwrap_or(0);
                    let target = (instr.codepoint as i64 + sign_extend(offset, 16)) as u32;
                    leaders.insert(target);
                    mark_next(&mut leaders);
                }
                // if-*z (zero compare)
                0x38..=0x3d => {
                    let offset = instr.v_b.unwrap_or(0);
                    let target = (instr.codepoint as i64 + sign_extend(offset, 16)) as u32;
                    leaders.insert(target);
                    mark_next(&mut leaders);
                }
                // switch
                0x2b | 0x2c => {
                    if let InstructionKind::Switch { ref table } = instr.kind {
                        for &rel in table.table.values() {
                            let target = (instr.codepoint as i64 + rel as i64) as u32;
                            leaders.insert(target);
                        }
                    }
                    mark_next(&mut leaders);
                }
                // return / throw
                0x0e..=0x11 | 0x27 => {
                    mark_next(&mut leaders);
                }
                _ => {}
            }
        }
        leaders
    }

    fn build_blocks(
        &self,
        leaders: &HashSet<u32>,
    ) -> (Vec<BasicBlock>, HashMap<u32, usize>, HashMap<usize, usize>) {
        let mut blocks: Vec<BasicBlock>         = Vec::new();
        let mut addr_lookup: HashMap<u32, usize> = HashMap::new();
        let mut instr_to_block: HashMap<usize, usize> = HashMap::new();

        let mut current_id: Option<usize> = None;

        for (idx, instr) in self.instructions.iter().enumerate() {
            if leaders.contains(&instr.codepoint) {
                let id = blocks.len();
                let mut block = BasicBlock::new(id);
                block.first_codepoint = instr.codepoint;
                blocks.push(block);
                addr_lookup.insert(instr.codepoint, id);
                current_id = Some(id);
            }
            if let Some(id) = current_id {
                blocks[id].instr_indices.push(idx);
                instr_to_block.insert(idx, id);
            }
        }

        (blocks, addr_lookup, instr_to_block)
    }
}

// ── Edge building ─────────────────────────────────────────────────────────────

fn classify_block(block: &mut BasicBlock, instrs: &[Instruction]) {
    let last_idx = *block.instr_indices.last().unwrap_or(&0);
    if block.instr_indices.is_empty() {
        return;
    }
    let last = &instrs[last_idx];
    let op = last.opcode;

    block.block_type = match op {
        0x0e..=0x11 => BasicBlockType::Return,
        0x27 => BasicBlockType::Throw,
        0x28 => {
            block.next_branch = Some((last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 8)) as u32);
            BasicBlockType::Goto
        }
        0x29 => {
            block.next_branch = Some((last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 16)) as u32);
            BasicBlockType::Goto
        }
        0x2a => {
            block.next_branch = Some((last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 32)) as u32);
            BasicBlockType::Goto
        }
        0x32..=0x37 => {
            block.next_branch = Some((last.codepoint as i64 + sign_extend(last.v_c.unwrap_or(0), 16)) as u32);
            BasicBlockType::If
        }
        0x38..=0x3d => {
            block.next_branch = Some((last.codepoint as i64 + sign_extend(last.v_b.unwrap_or(0), 16)) as u32);
            BasicBlockType::If
        }
        0x2b | 0x2c => BasicBlockType::Switch,
        _ => BasicBlockType::Generic,
    };
}

fn connect_edges(
    blocks: &mut Vec<BasicBlock>,
    edges: &mut Vec<CfgEdge>,
    addr_lookup: &HashMap<u32, usize>,
    sorted_cps: &[u32],
    instrs: &[Instruction],
) {
    let block_count = blocks.len();
    for block_id in 0..block_count {
        let last_instr_idx = *blocks[block_id].instr_indices.last().unwrap_or(&usize::MAX);
        if last_instr_idx == usize::MAX { continue; }
        let last = &instrs[last_instr_idx];
        let op = last.opcode;

        let next_block_id = |cp: u32| -> Option<usize> {
            let pos = sorted_cps.binary_search(&cp).ok()?;
            sorted_cps.get(pos + 1).and_then(|&next_cp| addr_lookup.get(&next_cp).copied())
        };

        let target_id = |cp: u32| -> Option<usize> { addr_lookup.get(&cp).copied() };

        let first_cp = blocks[block_id].first_codepoint;

        match op {
            0x28 => {
                let t = (last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 8)) as u32;
                if let Some(tid) = target_id(t) {
                    add_edge(blocks, edges, block_id, tid, EdgeKind::Jump, None);
                }
            }
            0x29 => {
                let t = (last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 16)) as u32;
                if let Some(tid) = target_id(t) {
                    add_edge(blocks, edges, block_id, tid, EdgeKind::Jump, None);
                }
            }
            0x2a => {
                let t = (last.codepoint as i64 + sign_extend(last.v_a.unwrap_or(0), 32)) as u32;
                if let Some(tid) = target_id(t) {
                    add_edge(blocks, edges, block_id, tid, EdgeKind::Jump, None);
                }
            }
            0x32..=0x37 => {
                let t = (last.codepoint as i64 + sign_extend(last.v_c.unwrap_or(0), 16)) as u32;
                if let Some(tid) = target_id(t) {
                    add_edge(blocks, edges, block_id, tid, EdgeKind::Jump, None);
                }
                if let Some(fid) = next_block_id(first_cp) {
                    add_edge(blocks, edges, block_id, fid, EdgeKind::FallThrough, None);
                }
            }
            0x38..=0x3d => {
                let t = (last.codepoint as i64 + sign_extend(last.v_b.unwrap_or(0), 16)) as u32;
                if let Some(tid) = target_id(t) {
                    add_edge(blocks, edges, block_id, tid, EdgeKind::Jump, None);
                }
                if let Some(fid) = next_block_id(first_cp) {
                    add_edge(blocks, edges, block_id, fid, EdgeKind::FallThrough, None);
                }
            }
            0x2b | 0x2c => {
                if let InstructionKind::Switch { ref table } = last.kind {
                    for (&key, &rel) in &table.table {
                        let t = (last.codepoint as i64 + rel as i64) as u32;
                        if let Some(tid) = target_id(t) {
                            add_edge(blocks, edges, block_id, tid, EdgeKind::Switch, Some(key));
                        }
                    }
                }
                if let Some(fid) = next_block_id(first_cp) {
                    add_edge(blocks, edges, block_id, fid, EdgeKind::FallThrough, None);
                }
            }
            0x0e..=0x11 | 0x27 => {
                // terminators — no successors
            }
            _ => {
                // fall-through
                if let Some(fid) = next_block_id(first_cp) {
                    add_edge(blocks, edges, block_id, fid, EdgeKind::FallThrough, None);
                }
            }
        }
    }
}

fn add_exception_edges(
    blocks: &mut Vec<BasicBlock>,
    edges: &mut Vec<CfgEdge>,
    addr_lookup: &HashMap<u32, usize>,
    try_items: &[super::parser::TryItem],
    handlers: &[super::parser::EncodedCatchHandler],
) {
    for try_item in try_items {
        let start = try_item.start_addr;
        let end   = start + try_item.insn_count as u32;

        // Every block whose first codepoint falls in [start, end) may throw
        let covered_blocks: Vec<usize> = addr_lookup
            .iter()
            .filter(|(&cp, _)| start <= cp && cp < end)
            .map(|(_, &id)| id)
            .collect();

        for handler in handlers {
            for h in &handler.handlers {
                let handler_addr = h.addr as u32;
                if let Some(&handler_block_id) = addr_lookup.get(&handler_addr) {
                    for &src_id in &covered_blocks {
                        add_edge(blocks, edges, src_id, handler_block_id, EdgeKind::Exception, None);
                    }
                }
            }
            if let Some(catch_all) = handler.catch_all_addr {
                let handler_addr = catch_all as u32;
                if let Some(&handler_block_id) = addr_lookup.get(&handler_addr) {
                    for &src_id in &covered_blocks {
                        add_edge(blocks, edges, src_id, handler_block_id, EdgeKind::Exception, None);
                    }
                }
            }
        }
    }
}

fn add_edge(
    blocks: &mut Vec<BasicBlock>,
    edges: &mut Vec<CfgEdge>,
    src_id: usize,
    dst_id: usize,
    kind: EdgeKind,
    switch_key: Option<i32>,
) {
    let edge_idx = edges.len();
    edges.push(CfgEdge { source_id: src_id, target_id: dst_id, kind, switch_key });
    blocks[src_id].successor_edges.push(edge_idx);
    blocks[dst_id].predecessor_edges.push(edge_idx);
}
